### PR TITLE
v2.5.0

### PR DESCRIPTION
**Pull request by arbrandes**
Introduces multi-provider support.

### CODE REVIEW

**Comment by fghaas** [comment]
Ah no worries, I won't. :)

**Comment by arbrandes** [comment]
Ok, this is ready for testing and code review.

File: README.md [hunk 1]
```diff
@@ -380,6 +377,17 @@ configured with the following attributes:
   shown as a dropdown to the user.  It must match the list of port numbers in
   both order and number.
 
+You can also use the following nested XML tags:
+
+* `provider`: (Optional) A list of OpenStack providers configured in the
```
Comment (fghaas) [inline README.md]: Minor nit: this sounds like the xblock expects a _single_ nested `<provider>` tag, which is then interpreted as a list. I'd suggest to rephrase this as:

```markdown
* `provider`: Reference to an OpenStack provider configured in the platform.
  It `name` attribute must match one of the providers in the XBlock configuration,
  `capacity` specifies how many environments should be launched in that
  provider at maximum (where 0 means keep launching environments until 
  encountering a launch failure), and 
  `environment` specifies a content store path to a Heat environment file specifying
  template parameters for that provider.
  This element must be specified at least once, and can be specified multiple times
  to use one or more fallback providers.
```

The part where it says "where 0 means launch until encountering a launch failure" is of course subject to the rest of the discussion, i.e. what does setting 0 mean, as opposed to not defining the attribute at all. Is it equivalent (meaning 0 means launch until failure), or is it 0 actually 0 (meaning launch nothing here at all, something that one might use to temporarily disable a provider in a deployed course) .

File: README.md [hunk 2]
```diff
@@ -380,6 +377,17 @@ configured with the following attributes:
   shown as a dropdown to the user.  It must match the list of port numbers in
   both order and number.
 
+You can also use the following nested XML tags:
```
Comment (fghaas) [inline README.md]: Suggest to use "elements" instead of "tags" here.

File: README.md [hunk 3]
```diff
@@ -380,6 +377,17 @@ configured with the following attributes:
...
+
+* `provider`: (Optional) A list of OpenStack providers configured in the
+  platform.  `name` must match one of the providers in the XBlock
+  configuration, `capacity` specifies how many simultaneous environments should
+  be launched in that provider before trying the next one, and `environment`
+  specifies a content store path to a Heat environment file specifying template
+  parameters for that provider.
+
+* `test`: (Optional) A list of scripts that will run against the lab
```
Comment (fghaas) [inline README.md]: Same as above. Suggest to clarify that you can define _multiple_ `<test>` elements which then _constitute_ a list of tests, rather than a _single_ `<test>` element that _contains_ a list of tests.

File: hastexo/hastexo.py [hunk 4]
```diff
@@ -126,6 +133,17 @@ def parse_xml(cls, node, runtime, keys, id_generator):
                 text = textwrap.dedent(text)
 
                 block.tests.append(text)
+            elif child.tag == "provider":
+                name = child.attrib["name"]
+                if not name:
+                    raise KeyError("name")
+                capacity = child.attrib.get("capacity", None)
```
Comment (fghaas) [inline hastexo/hastexo.py]: Since this defaults to `None`, what is the defined behavior if the user sets `0`, or a negative number?

File: hastexo/hastexo.py [hunk 5]
```diff
@@ -163,31 +181,25 @@ def get_block_ids(self):
...
             from xmodule.exceptions import NotFoundError
 
-            loc = StaticContent.compute_location(course_id,
-                                                 self.stack_template_path)
+            loc = StaticContent.compute_location(course_id, path)
             asset = contentstore().find(loc)
-            stack_template = asset.data
+            contents = asset.data
         except (ImportError, NotFoundError):
             pass
```
Comment (fghaas) [inline hastexo/hastexo.py]: Do we really want to succeed silently on failure here? Not even a log message if the course author fails to include a template, or botches the reference?

File: hastexo/hastexo.py [hunk 6]
```diff
@@ -318,24 +299,63 @@ def launch_stack_task_result(self, task_id):
...
         def _launch_stack(reset=False):
-            args = (
-                configuration,
-                self.stack_run,
-                self.stack_name,
-                self.get_stack_template(),
-                self.stack_user_name,
-                reset
-            )
+            stack_template = self.read_from_contentstore(
```
Comment (fghaas) [inline hastexo/hastexo.py]: What happens if this read fails? I see no exception handler here, which is generally fine — we do want to bail if the template can't be found. But would we get a stack trace somewhere?

File: hastexo/hastexo.py [hunk 7]
```diff
@@ -318,24 +299,63 @@ def launch_stack_task_result(self, task_id):
...
+                self.stack_template_path)
+            providers = []
+            if len(self.providers):
+                for provider in self.providers:
+                    p = dict(provider)
+                    env_path = p.get("environment")
+                    p["environment"] = self.read_from_contentstore(env_path)
```
Comment (fghaas) [inline hastexo/hastexo.py]: Same as above: how do we handle a read failure here?

File: hastexo/tasks.py [hunk 8]
```diff
@@ -76,174 +74,332 @@ def run(self,
...
+
+                stack_data = self.try_all_providers()
+            else:
+                stack_data = self.try_provider(stack.provider)
+        else:
+            stack_data = self.try_all_providers()
+
+        return stack_data
+
+    def get_provider_stack_count(self, provider):
```
Comment (fghaas) [inline hastexo/tasks.py]: I very much like the idea that instead of keeping track of a "stacks per provider" semaphore count, we just do a `SELECT COUNT` every time we need to get the stack count. We'll need to see how this scales, I suppose, but my hunch is that this scales far better than any contentious updates.

File: README.md (hunk 1 above)
Comment (arbrandes) [reply README.md]: +1 for improving the wording.  As for 0 versus `None`, they're currently equivalent, in that they translate to infinite capacity.  However, you have a point:  we may want to use -1 for infinite, and 0 for disabled.  What do you think?

File: hastexo/tasks.py (hunk 8 above)
Comment (arbrandes) [reply hastexo/tasks.py]: Yup, my feelings exactly.  However, I don't _think_ this will bog down our mariadb cluster any time soon.  The `hastexo_stack` table has not yet gone beyond 10k records in any of our production stacks, and unless we're talking tens of millions of stacks, we can probably handle our worst cases with some further InnoDB tuning.

**Comment by arbrandes** [comment]
I've addressed the comments in the latest fixup [commit](https://github.com/hastexo/hastexo-xblock/pull/79/commits/66eb7bad2c594ce064871bce4bbeb936d66f9893), which I'll squash later.

File: README.md [hunk 9]
```diff
@@ -377,17 +377,22 @@ configured with the following attributes:
...
+  Its `name` attribute must match one of the providers in the XBlock
+  configuration, `capacity` specifies how many environments should be launched
+  in that provider at maximum (where "-1" means keep launching environments
+  until encountering a launch failure, and "0" disables the provider), and
+  `environment` optionally specifies a content store path to a Heat environment
+  file specifying template parameters for that provider.  This element must be
+  specified at least once, and can be specified multiple times to use one or
+  more fallback providers.
+
+* `test`: The contents of this element will be run verbatum as a script in the
```
Comment (fghaas) [inline README.md]: It's "verbatim". :)

File: README.md [hunk 10]
```diff
@@ -377,17 +377,22 @@ configured with the following attributes:
...
+  user's lab environment, when the they click the "Check Progress" button.  As
+  such, it should define an interpreter via the "shabang" convention.  This
```
Comment (fghaas) [inline README.md]: Commonly "shebang".

File: hastexo/tasks.py [hunk 11]
```diff
@@ -175,12 +175,22 @@ def try_all_providers(self):
...
-
-            if capacity:
+            capacity = provider.get("capacity")
+            if capacity in (None, "None"):
+                capacity = -1
+            else:
+                try:
+                    capacity = int(capacity)
+                except (TypeError, ValueError):
+                    logger.info("Provider [%s] has invalid capacity." %
```
Comment (fghaas) [inline hastexo/tasks.py]: `.info()` seems weak here. This should be a warning or an error, I think.

**Comment by arbrandes** [comment]
All good points. :)  Addressed [here](https://github.com/hastexo/hastexo-xblock/pull/79/commits/584d6663e26c3981f4b6f191aec21a5a2a1ae0b0)

**Comment by arbrandes** [comment]
Any objections to merging this in?

**Comment by fghaas** [comment]
None, go ahead!
